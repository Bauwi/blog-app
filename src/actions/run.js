import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
