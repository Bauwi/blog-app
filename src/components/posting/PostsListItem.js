import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import moment from 'moment';

import KeywordsList from '../KeywordsList';

const PostsListItem = ({
  id, title, createdAt, author, stars, miniCover, keywords
}) => (
  <div className="list-item">
    <div className="list-item__content">
      <header>
        <img
          src={miniCover}
          className="list-item__img list-item__content__cover"
          alt="post cover"
        />
      </header>
      <section className="list-item__content__header">
        <div>
          <i className="fa fa-star" /> {stars}
        </div>
        <p className="list-item__content__date">{moment(createdAt).format('MMM Do, YYYY')}</p>
      </section>
      <div className="list-item__content__infos">
        <Link to={`/edit/${id}`}>
          <h3 className="list-item__content__title">{title}</h3>
        </Link>
        <div className="list-item__content__subtitle">
          <p className="list-item__content__author">{author}</p>
        </div>
      </div>
      <KeywordsList keywords={keywords} />
      <div className="list-item__content__category">
        <p className={`category border-${keywords.split(',')[0].trim()}`}>
          {keywords.split(',')[0].trim()}
        </p>
      </div>
    </div>
  </div>
);

PostsListItem.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  createdAt: PropTypes.number.isRequired,
  keywords: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  stars: PropTypes.number.isRequired,
  miniCover: PropTypes.string.isRequired
};

export default PostsListItem;
