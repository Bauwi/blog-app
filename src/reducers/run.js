const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
