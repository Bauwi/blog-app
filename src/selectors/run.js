const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
