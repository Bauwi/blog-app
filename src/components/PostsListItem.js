import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import moment from 'moment';

import { startAddStar } from '../actions/posts';

export class PostsListItem extends Component {
  onAddStar = () => {
    this.props.startAddStar(this.props.id, this.props.authorId, this.props.stars + 1);
  };

  render() {
    const { id, title, shortBody, keywords, createdAt } = this.props;
    const pathTo = this.props.isAuthenticated
      ? `/edit/${id}`
      : `/${this.props.authorId}/read/${id}`;
    return (
      <div>
        <Link to={pathTo}>
          <header>
            <h3>{title}</h3>
            <h5>Published {moment(createdAt).format('MMM Do, YYYY')}</h5>
          </header>
          <p>{shortBody}</p>
          <p>Keywords: {keywords}</p>
        </Link>
        <button onClick={this.onAddStar}>star</button>
      </div>
    );
  }
}

// const mapDispatchToProps = dispatch => ({
//   startAddStar: id => dispatch(startAddStar(id)),
//   startAddPost: post => dispatch(startAddPost(post))
// });

const mapDispatchToProps = dispatch => ({
  startAddStar: (id, authorId, count) => dispatch(startAddStar(id, authorId, count))
});

export default connect(undefined, mapDispatchToProps)(PostsListItem);
