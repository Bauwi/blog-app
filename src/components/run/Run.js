import React, { Component } from 'react';
import { connect } from 'react-redux';

import InlineList from '../InlineList';
import LoadingPage from '../LoadingPage';
import { startSetRunPosts } from '../../actions/run';

export class Run extends Component {
  state = {
    loading: true
  };
  componentWillMount() {
    if (this.props.run.length === 0) {
      this.props.startSetRunPosts().then(() => {
        this.setState(() => ({ loading: false }));
      });
    } else {
      this.setState(() => ({ loading: false }));
    }
  }

  render() {
    if (this.state.loading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <p>Run</p>
        <InlineList posts={this.props.run} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  run: state.run
});

export default connect(mapStateToProps, mapDispatchToProps)(Run);
