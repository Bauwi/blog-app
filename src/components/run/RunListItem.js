import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import moment from 'moment';

import KeywordsList from '../KeywordsList';

import { setCurrentPostRun } from '../../actions/run';

export class RunListItem extends Component {
  handleItemClick = () => {
    this.props.history.push('/run/start');
    this.props.setCurrentPostRun(this.props.post.content.id);
  };

  render() {
    const {
      id,
      title,
      createdAt,
      author,
      keywords,
      authorId,
      miniCover,
      stars
    } = this.props.post.content;
    return (
      <div>
        <div className="inline-list-item">
          <header className="inline-list-item__left">
            <img src={miniCover} className="inline-list-item__cover" alt="post cover" />
          </header>
          <div className="inline-list-item__right">
            <section className="inline-list-item__header">
              <div>
                <i className="fa fa-star" /> {stars}
              </div>
              <div className="inline-list-item__header--right">
                {this.props.state === 'read' && (
                  <p>
                    <i className="fa fa-check" />
                  </p>
                )}
                <p className="inline-list-item__date">{moment(createdAt).format('MMM Do, YYYY')}</p>
              </div>
            </section>
            <div onClick={this.handleItemClick} className="inline-list-item__infos">
              <h3 className="inline-list-item__title">{title}</h3>
            </div>
            <footer className="inline-list-item__footer">
              <p className="inline-list-item__author">{author}</p>
              <KeywordsList keywords={keywords} />
            </footer>
          </div>
        </div>
      </div>
    );
  }
}

RunListItem.propTypes = {
  post: PropTypes.object.isRequired,
  state: PropTypes.string.isRequired,
  setCurrentPostRun: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  setCurrentPostRun: id => dispatch(setCurrentPostRun(id))
});

export default withRouter(connect(undefined, mapDispatchToProps)(RunListItem));
