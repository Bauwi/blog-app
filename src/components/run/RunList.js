import React, { Component } from 'react';
import PropTypes from 'prop-types';

import RunListItem from './RunListItem';

export default class RunList extends Component {
  renderList() {
    return this.props.posts.map((post) => {
      const key = post.content ? post.content.id : post.id;

      return <RunListItem key={key} post={post} state={post.state} />;
    });
  }

  render() {
    return (
      <div className="content-container run-list">
        {this.props.posts && this.props.posts.length !== 0 ? (
          this.renderList()
        ) : (
          <div className="run-list__empty">
            <p>Add a post to your run to start a session</p>
          </div>
        )}
      </div>
    );
  }
}

RunList.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object)
};

RunList.defaultProps = {
  posts: []
};
