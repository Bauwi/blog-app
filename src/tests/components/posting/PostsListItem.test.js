import React from 'react';
import { shallow } from 'enzyme';
import PostsListItem from '../../../components/posting/PostsListItem';
import run from '../../fixtures/run';

test('should render PostsListItem component properly', () => {
  const wrapper = shallow(<PostsListItem {...run.posts[1].content} />);
  expect(wrapper).toMatchSnapshot();
});
