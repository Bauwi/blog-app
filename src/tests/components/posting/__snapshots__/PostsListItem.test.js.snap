// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostsListItem component properly 1`] = `
<div
  className="list-item"
>
  <Link
    replace={false}
    to="/edit/2"
  >
    <div
      className="list-item__content"
    >
      <header>
        <img
          alt="post cover"
          className="list-item__img list-item__content__cover"
          src="https://www.w3schools.com/w3css/img_fjords.jpg"
        />
      </header>
      <section
        className="list-item__content__header"
      >
        <div>
          <i
            className="fa fa-star"
          />
           
          25
        </div>
        <p
          className="list-item__content__date"
        >
          Jan 1st, 1970
        </p>
      </section>
      <div
        className="list-item__content__infos"
      >
        <h3
          className="list-item__content__title"
        >
          Post 2 in Run
        </h3>
        <div
          className="list-item__content__subtitle"
        >
          <p
            className="list-item__content__author"
          >
            author2
          </p>
        </div>
      </div>
      <KeywordsList
        keywords="cat, dog, fish"
      />
    </div>
  </Link>
</div>
`;
